### DISCUSSION TITLE
Proposal: fork pvfactors to a new pvlib project

### DISCUSSION THREAD

**Discussion by kandersolar** [pvlib/pvlib-python#1657] [General]
## Background

pvfactors is an open-source python package originally developed by SunPower.  It implements a 2-D view factors model for determining front- and rear-side irradiance for bifacial PV systems, considering things like row-row and row-ground shadows.  It is slower than `pvlib.infinite_sheds` but can model some things that `infinite_sheds` cannot, for example finite numbers of rows and irradiance at multiple locations along the row's slant height.  pvfactors can be used directly or via `pvlib.bifacial.pvfactors_timeseries`, a convenience function that wraps pvfactors as an optional dependency.  Its PVSC paper has been cited [30+ times](https://scholar.google.com/scholar?cluster=17641826033504233184&hl=en&as_sdt=4005&sciodt=0,6) since 2017 and I can attest to ongoing use of pvfactors in PV industry.

Aside from the code itself, pvfactors has sphinx-based [documentation](https://sunpower.github.io/pvfactors/) and a pytest-based [test suite](https://github.com/SunPower/pvfactors/tree/master/pvfactors/tests), both of which will feel familiar to those who work on pvlib-python.  The pvfactors repository contents are BSD-3 licensed.

It was developed and maintained by @anomam until he left SunPower a few years ago.  Since then, the project has been mostly inactive, with just a few PRs getting [merged](https://github.com/SunPower/pvfactors/graphs/contributors) between @anomam, @markcampanelli, and me.  The main road block is the lack of overlap between the people who are motivated to work on the project and the people with the admin privileges necessary to properly manage the project.  Unless something changes on the SunPower side, I'm not confident that the package will receive the maintenance it needs to continue working as the rest of the python ecosystem evolves.  Over a year ago we first approached SunPower regarding the future of pvfactors: https://github.com/SunPower/pvfactors/issues/127.  Unfortunately that conversation hasn't reached any satisfying conclusion.  In the meantime, bug reports and dependency incompatibilities accumulate.  

## Proposal

In the absence of SunPower actively maintaining pvfactors or agreeing to transfer ownership to another steward, I propose we fork the pvfactors repository into a new repository in the pvlib GitHub organization, release it as a new package on PyPI, and have pvlib's wrapper function use our fork.  Specifically, I am thinking of a sequence like this:

1. Fork the `SunPower/pvfactors` repository to `pvlib/pvfactors`, and then rename the repository to something other than `pvfactors`.  Renaming is necessary so that we can eventually publish it to PyPI without a name conflict.
2. Migrate the CI jobs to GitHub Actions and the sphinx docs to ReadTheDocs to parallel the setup for our other repositories.
3. Fix the most pressing issue: incompatibility with Shapely 2.0 (https://github.com/SunPower/pvfactors/pull/130).
4. Set up automated releases like we have for other repos and make an initial release to PyPI under the new project name.
5. Swap pvlib's dependency and code over to the new package.  It might make sense to rename the pvlib function as well (since it has `pvfactors` in the name), in which case we should consider making other changes/improvements while we're at it.
6. (longer term) address previous bug reports, improve docs, etc.  I have some additional ideas for speed and capability enhancements that I would like to pursue.

I am interested and willing to be a maintainer of the new repository.  @anomam has expressed interest as well although may not have much time available to contribute directly.  The involvement of other maintainers would be very welcome but my support for this plan is not contingent on it.

## What is needed from you

1. I would like the rest of the pvlib "core team" to indicate agreement (or at least a lack of objection) to this plan.
2. Suggestions for the new repository/package name.  Perhaps something with "segments" (a central concept in pvfactors)?

**Comment by mikofski** [comment]
I don't think it's necessary to rename the pvlib function - I think the PyPI name is independent from the namespace - I think we should continue to reuse the same "pvfactors" namespace in pvlib and which pvfactors gets used will depend on what the user installs

**Reply by kandersolar** [reply]
I think there are three main places where the name is relevant: PyPI (`pip install pvfactors`), the package namespace (`import pvfactors`), and pvlib's API (`pvlib.bifacial.pvfactors_timeseries`).  The first I think _must_ be renamed since we cannot commandeer that package name on PyPI.  It sounds like you are advocating for not changing the latter two. I agree that there is no technical reason forcing our hand for them, but keeping the pvfactors name at all seems like a footgun to me.

Pros of renaming:
- We aren't setting ourselves up to explain "actually you need to install the _other_ pvfactors for X to work" as our repo diverges from the original over time
- Avoiding confusion where people go looking at the wrong source code/reporting bugs in the wrong repo because they don't realize that our "pvfactors" doesn't come from `SunPower/pvfactors`

Pros of keeping "pvfactors":
- Convenience for us and other users if our "pvfactors" is a transparent drop-in replacement for the original
- No annoying deprecation cycle for the pvlib function
- Name recognition; "pvfactors" is a household name (in some circles)

**Reply by wholmgren** [reply]
No need to worry about deprecation if the default user experience is already broken. 

Thinking of a name that is similar enough to spark some recognition while different enough that I won't mistype it.... "pvlib-factors" and "pvviewfactors" come to mind.

**Reply by AdamRJensen** [reply]
`pvviewfactors` sounds neat!

I still think it's a bit of a shame not transferring the existing repo (and thus keeping the convenient GH issue and PR history), but alas if SunPower is not willing or able, I suppose there is not much we can do.

**Reply by kandersolar** [reply]
> No need to worry about deprecation if the default user experience is already broken.

It's not broken quite yet (e.g. our test suite still works), but I'm guessing its current `shapely<2.0` requirement will become an issue when python 3.12 is released later this year.  I'm pushing this fork primarily as a proactive measure against future breakage.

**Reply by wholmgren** [reply]
> I still think it's a bit of a shame not transferring the existing repo (and thus keeping the convenient GH issue and PR history), but alas if SunPower is not willing or able, I suppose there is not much we can do.

I agree, but I think we're well past the time to move on.

**Reply by kandersolar** [reply]
FYI: I had some time this weekend and forked SunPower/pvfactors to [pvlib/solarfactors](https://github.com/pvlib/solarfactors) so I could get started on the list of steps outlined above.  I'm open to changing the name to something else; I just didn't want to let the name decision prevent progression on the other items.

**Reply by kandersolar** [reply]
And to wrap up this thread: motivated by #1796, pvlib's dependency was switched from SunPower/pvfactors to pvlib/solarfactors in #1797.
